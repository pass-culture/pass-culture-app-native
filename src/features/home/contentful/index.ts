export * from './contentful'
export { ExclusivityPane, BusinessPane, Offers, OffersWithCover, VenuesModule } from './moduleTypes'
export { processHomepageEntry } from './processHomepageEntry'
