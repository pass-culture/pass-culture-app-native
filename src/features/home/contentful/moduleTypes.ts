import {
  ExclusivityDisplayParametersFields,
  DisplayParametersFields,
  SearchParametersFields,
  VenuesSearchParametersFields,
} from './contentful'

export class Offers {
  search: SearchParametersFields[]
  display: DisplayParametersFields
  moduleId: string
  constructor({
    search,
    display,
    moduleId,
  }: {
    search: SearchParametersFields[]
    display: DisplayParametersFields
    moduleId: string
  }) {
    this.search = search
    this.display = display
    this.moduleId = moduleId
  }
}

export class OffersWithCover extends Offers {
  cover: string | null
  constructor({
    cover,
    display,
    search,
    moduleId,
  }: {
    search: SearchParametersFields[]
    display: DisplayParametersFields
    cover: string | null
    moduleId: string
  }) {
    super({ search, display, moduleId })
    this.cover = cover
  }
}

export class VenuesModule {
  search: VenuesSearchParametersFields[]
  display: DisplayParametersFields
  moduleId: string
  constructor({
    search,
    display,
    moduleId,
  }: {
    search: VenuesSearchParametersFields[]
    display: DisplayParametersFields
    moduleId: string
  }) {
    this.search = search
    this.display = display
    this.moduleId = moduleId
  }
}

export class ExclusivityPane {
  alt: string
  image: string
  id: number
  moduleId: string
  display?: ExclusivityDisplayParametersFields
  constructor({
    alt,
    image,
    id,
    moduleId,
    display,
  }: {
    alt: string
    image: string
    id: number
    moduleId: string
    display?: ExclusivityDisplayParametersFields
  }) {
    this.alt = alt
    this.image = image
    this.id = id
    this.moduleId = moduleId
    this.display = display
  }
}

export class RecommendationPane {
  display: DisplayParametersFields
  constructor({ display }: { display: DisplayParametersFields }) {
    this.display = display
  }
}

export class BusinessPane {
  title: string
  firstLine: string | undefined
  image: string
  leftIcon: string | undefined
  secondLine: string | undefined
  url: string | undefined
  moduleId: string
  targetNotConnectedUsersOnly: boolean | undefined
  constructor({
    title,
    firstLine,
    image,
    leftIcon,
    secondLine,
    url,
    moduleId,
    targetNotConnectedUsersOnly,
  }: {
    title: string
    firstLine: string | undefined
    image: string
    leftIcon: string | undefined
    secondLine: string | undefined
    url: string | undefined
    moduleId: string
    targetNotConnectedUsersOnly: boolean | undefined
  }) {
    this.title = title
    this.firstLine = firstLine
    this.image = image
    this.leftIcon = leftIcon
    this.secondLine = secondLine
    this.url = url
    this.moduleId = moduleId
    this.targetNotConnectedUsersOnly = targetNotConnectedUsersOnly
  }
}

export type ProcessedModule =
  | Offers
  | ExclusivityPane
  | BusinessPane
  | OffersWithCover
  | RecommendationPane
  | VenuesModule
