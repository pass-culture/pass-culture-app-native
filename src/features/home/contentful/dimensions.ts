import { Layout } from 'features/home/contentful'
import { LENGTH_L, LENGTH_M, RATIO_HOME_IMAGE } from 'ui/theme'

export function getPlaylistItemDimensionsFromLayout(layout: Layout): {
  itemWidth: number
  itemHeight: number
} {
  const itemHeight = layout === 'two-items' ? LENGTH_M : LENGTH_L
  const itemWidth = itemHeight * RATIO_HOME_IMAGE
  return { itemWidth, itemHeight }
}
