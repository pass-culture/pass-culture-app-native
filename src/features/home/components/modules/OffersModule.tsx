import React, { useCallback, useEffect } from 'react'

import { useAuthContext } from 'features/auth/context/AuthContext'
import { HomeOfferTile } from 'features/home/components/HomeOfferTile'
import { ModuleData, OffersModule as OffersModuleType } from 'features/home/types'
import { getSearchStackConfig } from 'features/navigation/SearchStackNavigator/helpers'
import { useAdaptOffersPlaylistParameters } from 'libs/algolia/fetchAlgolia/fetchMultipleOffers/helpers/useAdaptOffersPlaylistParameters'
import { analytics } from 'libs/analytics'
import { ContentTypes } from 'libs/contentful/types'
import { usePlaylistItemDimensionsFromLayout } from 'libs/contentful/usePlaylistItemDimensionsFromLayout'
import { useFeatureFlag } from 'libs/firebase/firestore/featureFlags/useFeatureFlag'
import { RemoteStoreFeatureFlags } from 'libs/firebase/firestore/types'
import useFunctionOnce from 'libs/hooks/useFunctionOnce'
import { formatDates } from 'libs/parsers/formatDates'
import { getDisplayPrice } from 'libs/parsers/getDisplayPrice'
import { useCategoryHomeLabelMapping, useCategoryIdMapping } from 'libs/subcategories'
import { Offer } from 'shared/offer/types'
import { PassPlaylist } from 'ui/components/PassPlaylist'
import { CustomListRenderItem, ItemDimensions, RenderFooterItem } from 'ui/components/Playlist'
import { SeeMore } from 'ui/components/SeeMore'

export type OffersModuleProps = {
  offersModuleParameters: OffersModuleType['offersModuleParameters']
  displayParameters: OffersModuleType['displayParameters']
  moduleId: string
  index: number
  homeEntryId: string | undefined
  data: ModuleData | undefined
}

const keyExtractor = (item: Offer) => item.objectID

export const OffersModule = (props: OffersModuleProps) => {
  const isNewOfferTileDisplayed = useFeatureFlag(RemoteStoreFeatureFlags.WIP_NEW_OFFER_TILE)
  const { displayParameters, offersModuleParameters, index, moduleId, homeEntryId, data } = props
  const adaptedPlaylistParameters = useAdaptOffersPlaylistParameters()
  const mapping = useCategoryIdMapping()
  const labelMapping = useCategoryHomeLabelMapping()
  const { user } = useAuthContext()

  const { playlistItems, nbPlaylistResults } = data ?? {
    playlistItems: [],
    nbPlaylistResults: 0,
  }

  const [parameters] = offersModuleParameters
  // When we navigate to the search page, we want to show 20 results per page,
  // not what is configured in contentful
  // eslint-disable-next-line react-hooks/exhaustive-deps

  // @ts-expect-error: because of noUncheckedIndexedAccess
  const { offerParams, locationParams } = adaptedPlaylistParameters(parameters)
  const searchParams = {
    ...offerParams,
    locationParams,
    hitsPerPage: 20,
  }
  const searchTabConfig = getSearchStackConfig('SearchResults', searchParams)
  // @ts-expect-error: because of noUncheckedIndexedAccess
  const moduleName = displayParameters.title ?? parameters.title
  const logHasSeenAllTilesOnce = useFunctionOnce(() =>
    analytics.logAllTilesSeen({ moduleName, numberOfTiles: playlistItems.length })
  )

  const showSeeMore =
    nbPlaylistResults &&
    playlistItems.length < nbPlaylistResults &&
    // @ts-expect-error: because of noUncheckedIndexedAccess
    !(parameters.tags ?? parameters.beginningDatetime ?? parameters.endingDatetime)

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const onPressSeeMore = showSeeMore
    ? () => {
        analytics.logClickSeeMore({ moduleName, moduleId })
      }
    : undefined

  const renderItem: CustomListRenderItem<Offer> = useCallback(
    ({ item, width, height }) => {
      const timestampsInMillis = item.offer.dates?.map((timestampInSec) => timestampInSec * 1000)
      return (
        <HomeOfferTile
          offerLocation={item._geoloc}
          categoryLabel={labelMapping[item.offer.subcategoryId]}
          categoryId={mapping[item.offer.subcategoryId]}
          subcategoryId={item.offer.subcategoryId}
          offerId={+item.objectID}
          name={item.offer.name}
          date={formatDates(timestampsInMillis)}
          isDuo={item.offer.isDuo}
          thumbUrl={item.offer.thumbUrl}
          price={getDisplayPrice(item.offer.prices)}
          isBeneficiary={user?.isBeneficiary}
          moduleName={moduleName}
          moduleId={moduleId}
          homeEntryId={homeEntryId}
          width={width}
          height={height}
          variant={isNewOfferTileDisplayed ? 'new' : 'default'}
        />
      )
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [user?.isBeneficiary, labelMapping, mapping]
  )

  const { itemWidth, itemHeight } = usePlaylistItemDimensionsFromLayout(
    displayParameters.layout,
    homeEntryId
  )

  const renderFooter: RenderFooterItem = useCallback(
    ({ width, height }: ItemDimensions) => {
      return showSeeMore ? (
        <SeeMore
          navigateTo={{ screen: searchTabConfig[0], params: searchTabConfig[1], withPush: true }}
          width={width}
          height={height}
          onPress={onPressSeeMore as () => void}
        />
      ) : null
    },
    [onPressSeeMore, showSeeMore, searchTabConfig]
  )

  const shouldModuleBeDisplayed =
    playlistItems.length > 0 && playlistItems.length >= displayParameters.minOffers

  useEffect(() => {
    if (shouldModuleBeDisplayed) {
      analytics.logModuleDisplayedOnHomepage({
        moduleId,
        moduleType: ContentTypes.ALGOLIA,
        index,
        homeEntryId,
        offers: (playlistItems as Offer[]).map((item) => item.objectID),
      })
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [shouldModuleBeDisplayed])

  if (!shouldModuleBeDisplayed) return null

  return (
    <PassPlaylist
      testID="offersModuleList"
      title={displayParameters.title}
      subtitle={displayParameters.subtitle}
      data={playlistItems}
      itemHeight={itemHeight}
      itemWidth={itemWidth}
      onPressSeeMore={onPressSeeMore}
      titleSeeMoreLink={{ screen: searchTabConfig[0], params: searchTabConfig[1] }}
      renderItem={renderItem}
      renderFooter={renderFooter}
      keyExtractor={keyExtractor}
      onEndReached={logHasSeenAllTilesOnce}
    />
  )
}
