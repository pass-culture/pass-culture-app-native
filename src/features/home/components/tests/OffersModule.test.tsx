import { render, act } from '@testing-library/react-native'
import mockdate from 'mockdate'
import React from 'react'

import { mockedAlgoliaResponse } from 'libs/algolia/mockedResponses/mockedAlgoliaResponse'
import { logAllTilesSeen } from 'libs/analytics'
import { reactQueryProviderHOC } from 'tests/reactQueryProviderHOC'
import { flushAllPromises } from 'tests/utils'
import { ColorsEnum } from 'ui/theme'

import { AlgoliaParametersFields, DisplayParametersFields } from '../../contentful/contentful'
import { OffersModule } from '../OffersModule'

mockdate.set(new Date(2020, 10, 16))

const props = {
  algolia: {} as AlgoliaParametersFields,
  display: {
    minOffers: 0,
    title: 'Module title',
    layout: 'one-item-medium',
  } as DisplayParametersFields,
  hits: mockedAlgoliaResponse.hits,
  nbHits: mockedAlgoliaResponse.nbHits,
  cover: null,
  position: null,
}

describe('OffersModule component', () => {
  it('should render correctly - with black title', () => {
    const component = render(reactQueryProviderHOC(<OffersModule {...props} index={1} />))
    expect(component).toMatchSnapshot()
    expect(component.getByTestId('moduleTitle').props.color).toBe(ColorsEnum.BLACK)
  })
  it('should render with white title if firert module displayed', async () => {
    const component = render(reactQueryProviderHOC(<OffersModule {...props} index={0} />))
    expect(component.getByTestId('moduleTitle').props.color).toBe(ColorsEnum.WHITE)
  })
})

describe('OffersModule component - Analytics', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })
  it('should trigger logEvent "AllTilesSeen" when reaching the end of the module', async () => {
    const component = render(reactQueryProviderHOC(<OffersModule {...props} index={1} />))
    const flatList = component.getByTestId('offersModuleList')

    await act(async () => {
      await flatList.props.onEndReached()
      await flushAllPromises()
    })

    expect(logAllTilesSeen).toHaveBeenCalledWith(props.display.title, props.nbHits)
  })

  it('should trigger logEvent "AllTilesSeen" only once', async () => {
    const component = render(reactQueryProviderHOC(<OffersModule {...props} index={1} />))
    const flatList = component.getByTestId('offersModuleList')

    await act(async () => {
      // 1st scroll to last item => trigger
      await flatList.props.onEndReached()
      await flushAllPromises()
    })
    expect(logAllTilesSeen).toHaveBeenCalledWith(props.display.title, props.nbHits)
    expect(logAllTilesSeen).toHaveBeenCalledTimes(1)

    flatList.props.onEndReached()
    expect(logAllTilesSeen).toHaveBeenCalledTimes(1)
  })
})
