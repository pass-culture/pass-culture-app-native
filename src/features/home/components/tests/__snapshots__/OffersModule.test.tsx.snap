// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OffersModule component should render correctly - with black title 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "marginHorizontal": 24,
        },
      ]
    }
  >
    <Text
      color="#151515"
      numberOfLines={2}
      style={
        Array [
          Object {
            "color": "#151515",
            "fontFamily": "Montserrat-Bold",
            "fontSize": 20,
            "lineHeight": 24,
          },
        ]
      }
      testID="moduleTitle"
    >
      Module title
    </Text>
  </View>,
  <View
    numberOfSpaces={4}
    style={
      Array [
        Object {
          "height": 16,
        },
      ]
    }
  />,
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListFooterComponent={
      <Styled(View)
        numberOfSpaces={6}
      />
    }
    ListHeaderComponent={
      <Styled(View)
        numberOfSpaces={6}
      />
    }
    data={
      Array [
        Object {
          "_geoloc": Object {
            "lat": 48.94374,
            "lng": 2.48171,
          },
          "objectID": "AGHYQ",
          "offer": Object {
            "author": null,
            "category": "MUSIQUE",
            "dateCreated": 1604358109.991316,
            "dates": Array [],
            "description": "Bel astre voyageur, hôte qui nous arrives, Des profondeurs du ciel et qu'on n'attendait pas, Où vas-tu ? Quel dessein pousse vers nous tes pas ? Toi qui vogues au large en cette mer sans rives",
            "id": "AGHYQ",
            "isDigital": false,
            "isDuo": false,
            "isEvent": false,
            "isThing": true,
            "isbn": null,
            "label": "Abonnements concerts",
            "musicSubType": "914",
            "musicType": "900",
            "name": "Mensch ! Où sont les Hommes ?",
            "performer": null,
            "priceMax": 28,
            "priceMin": 28,
            "prices": Array [
              28,
            ],
            "showSubType": null,
            "showType": null,
            "speaker": null,
            "stageDirector": null,
            "stocksDateCreated": Array [
              1604358115.556418,
            ],
            "tags": Array [],
            "thumbUrl": "https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CDNQ",
            "times": Array [],
            "type": "Écouter",
            "visa": null,
            "withdrawalDetails": null,
          },
          "offerer": Object {
            "name": "Danse Jazz Dans Tes Bottes",
          },
          "venue": Object {
            "city": "Aulnay-sous-Bois",
            "departementCode": "93",
            "name": "Michel et son accordéon",
            "publicName": null,
          },
        },
        Object {
          "_geoloc": Object {
            "lat": 48.91265,
            "lng": 2.4513,
          },
          "objectID": "AGHYA",
          "offer": Object {
            "author": "Catherine Mater",
            "category": "MUSIQUE",
            "dateCreated": 1604358109.991316,
            "dates": Array [],
            "description": "D'un coup d'épée, Frappé par un héros, tomber la pointe au coeur! Oui, je disais cela!... Le destin est railleur!... Et voila que je suis tué, par un laquais, d'un coup de bûche! C'est très bien. J'aurai tout manqué, même ma mort.",
            "id": "AGHYA",
            "isDigital": false,
            "isDuo": false,
            "isEvent": false,
            "isThing": true,
            "isbn": null,
            "label": "Musique",
            "musicSubType": "914",
            "musicType": "900",
            "name": "I want something more",
            "performer": "Frank Despacito",
            "priceMax": 23,
            "priceMin": 23,
            "prices": Array [
              23,
            ],
            "showSubType": null,
            "showType": null,
            "speaker": null,
            "stageDirector": null,
            "stocksDateCreated": Array [
              1604358115.524528,
            ],
            "tags": Array [],
            "thumbUrl": "https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CDKQ",
            "times": Array [],
            "type": "Écouter",
            "visa": null,
            "withdrawalDetails": null,
          },
          "offerer": Object {
            "name": "Club Dorothy",
          },
          "venue": Object {
            "city": "Drancy",
            "departementCode": "93",
            "name": "Maison de la Brique",
            "publicName": null,
          },
        },
        Object {
          "_geoloc": Object {
            "lat": 4.90339,
            "lng": -52.31663,
          },
          "objectID": "AGHWS",
          "offer": Object {
            "author": null,
            "category": "MUSIQUE",
            "dateCreated": 1604358109.991255,
            "dates": Array [
              1605643200,
            ],
            "description": null,
            "id": "AGHWS",
            "isDigital": false,
            "isDuo": true,
            "isEvent": true,
            "isThing": false,
            "isbn": null,
            "label": "Concert ou festival",
            "musicSubType": null,
            "musicType": null,
            "name": "Un lit sous une rivière",
            "performer": null,
            "priceMax": 34,
            "priceMin": 34,
            "prices": Array [
              34,
            ],
            "showSubType": null,
            "showType": null,
            "speaker": null,
            "stageDirector": null,
            "stocksDateCreated": Array [
              1604358115.242088,
            ],
            "tags": Array [],
            "thumbUrl": "https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CDBA",
            "times": Array [
              72000,
            ],
            "type": "Écouter",
            "visa": null,
            "withdrawalDetails": null,
          },
          "offerer": Object {
            "name": "Théâtre Balboa",
          },
          "venue": Object {
            "city": "Cayenne",
            "departementCode": "973",
            "name": "Espace des Gnoux",
            "publicName": null,
          },
        },
        Object {
          "_geoloc": Object {
            "lat": 4.90339,
            "lng": -52.31663,
          },
          "objectID": "AGH2M",
          "offer": Object {
            "author": null,
            "category": "MUSIQUE",
            "dateCreated": 1604358109.991316,
            "dates": Array [],
            "description": "D'un coup d'épée, Frappé par un héros, tomber la pointe au coeur! Oui, je disais cela!... Le destin est railleur!... Et voila que je suis tué, par un laquais, d'un coup de bûche! C'est très bien. J'aurai tout manqué, même ma mort.",
            "id": "AGH2M",
            "isDigital": false,
            "isDuo": false,
            "isEvent": false,
            "isThing": true,
            "isbn": null,
            "label": "Abonnements concerts",
            "musicSubType": "892",
            "musicType": "880",
            "name": "I want something more",
            "performer": null,
            "priceMax": 28,
            "priceMin": 28,
            "prices": Array [
              28,
            ],
            "showSubType": null,
            "showType": null,
            "speaker": null,
            "stageDirector": null,
            "stocksDateCreated": Array [
              1604358115.646641,
            ],
            "tags": Array [],
            "thumbUrl": "https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CDZQ",
            "times": Array [],
            "type": "Écouter",
            "visa": null,
            "withdrawalDetails": null,
          },
          "offerer": Object {
            "name": "Théâtre Balboa",
          },
          "venue": Object {
            "city": "Cayenne",
            "departementCode": "973",
            "name": "Espace des Gnoux",
            "publicName": null,
          },
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.1}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    testID="offersModuleList"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          numberOfSpaces={6}
          style={
            Array [
              Object {
                "width": 24,
              },
            ]
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 8,
                  "height": 312,
                },
              ]
            }
          >
            <View>
              <Image
                imageHeight={288}
                imageWidth={192}
                source={
                  Object {
                    "uri": "https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CDNQ",
                  }
                }
                style={
                  Array [
                    Object {
                      "borderTopLeftRadius": 8,
                      "borderTopRightRadius": 8,
                      "height": 288,
                      "width": 192,
                    },
                  ]
                }
                testID="offerTileImage"
              />
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#151515",
                      "borderBottomLeftRadius": 8,
                      "borderBottomRightRadius": 8,
                      "flexDirection": "row",
                      "height": 24,
                      "width": 192,
                    },
                  ]
                }
                width={192}
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexBasis": 0,
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "justifyContent": "center",
                        "paddingHorizontal": 4,
                      },
                    ]
                  }
                >
                  <Text
                    color="#ffffff"
                    style={
                      Array [
                        Object {
                          "color": "#ffffff",
                          "fontFamily": "Montserrat-SemiBold",
                          "fontSize": 12,
                          "lineHeight": 16,
                        },
                      ]
                    }
                    testID="categoryImageCaption"
                  >
                    Musique
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            imageWidth={192}
            style={
              Array [
                Object {
                  "marginTop": 8,
                  "maxWidth": 192,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#151515",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
            >
              Mensch ! Où sont les Hommes ?
            </Text>
            <Text
              color="#626262"
              style={
                Array [
                  Object {
                    "color": "#626262",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
              testID="priceIsDuo"
            >
              28 €
            </Text>
          </View>
        </View>
        <View
          numberOfSpaces={4}
          style={
            Array [
              Object {
                "width": 16,
              },
            ]
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 8,
                  "height": 312,
                },
              ]
            }
          >
            <View>
              <Image
                imageHeight={288}
                imageWidth={192}
                source={
                  Object {
                    "uri": "https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CDKQ",
                  }
                }
                style={
                  Array [
                    Object {
                      "borderTopLeftRadius": 8,
                      "borderTopRightRadius": 8,
                      "height": 288,
                      "width": 192,
                    },
                  ]
                }
                testID="offerTileImage"
              />
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#151515",
                      "borderBottomLeftRadius": 8,
                      "borderBottomRightRadius": 8,
                      "flexDirection": "row",
                      "height": 24,
                      "width": 192,
                    },
                  ]
                }
                width={192}
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexBasis": 0,
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "justifyContent": "center",
                        "paddingHorizontal": 4,
                      },
                    ]
                  }
                >
                  <Text
                    color="#ffffff"
                    style={
                      Array [
                        Object {
                          "color": "#ffffff",
                          "fontFamily": "Montserrat-SemiBold",
                          "fontSize": 12,
                          "lineHeight": 16,
                        },
                      ]
                    }
                    testID="categoryImageCaption"
                  >
                    Musique
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            imageWidth={192}
            style={
              Array [
                Object {
                  "marginTop": 8,
                  "maxWidth": 192,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#151515",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
            >
              I want something more
            </Text>
            <Text
              color="#626262"
              style={
                Array [
                  Object {
                    "color": "#626262",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
              testID="priceIsDuo"
            >
              23 €
            </Text>
          </View>
        </View>
        <View
          numberOfSpaces={4}
          style={
            Array [
              Object {
                "width": 16,
              },
            ]
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 8,
                  "height": 312,
                },
              ]
            }
          >
            <View>
              <Image
                imageHeight={288}
                imageWidth={192}
                source={
                  Object {
                    "uri": "https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CDBA",
                  }
                }
                style={
                  Array [
                    Object {
                      "borderTopLeftRadius": 8,
                      "borderTopRightRadius": 8,
                      "height": 288,
                      "width": 192,
                    },
                  ]
                }
                testID="offerTileImage"
              />
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#151515",
                      "borderBottomLeftRadius": 8,
                      "borderBottomRightRadius": 8,
                      "flexDirection": "row",
                      "height": 24,
                      "width": 192,
                    },
                  ]
                }
                width={192}
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexBasis": 0,
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "justifyContent": "center",
                        "paddingHorizontal": 4,
                      },
                    ]
                  }
                >
                  <Text
                    color="#ffffff"
                    style={
                      Array [
                        Object {
                          "color": "#ffffff",
                          "fontFamily": "Montserrat-SemiBold",
                          "fontSize": 12,
                          "lineHeight": 16,
                        },
                      ]
                    }
                    testID="categoryImageCaption"
                  >
                    Musique
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            imageWidth={192}
            style={
              Array [
                Object {
                  "marginTop": 8,
                  "maxWidth": 192,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#151515",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
            >
              Un lit sous une rivière
            </Text>
            <Text
              color="#626262"
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#626262",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
            >
              17 novembre 2020
            </Text>
            <Text
              color="#626262"
              style={
                Array [
                  Object {
                    "color": "#626262",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
              testID="priceIsDuo"
            >
              34 €
            </Text>
          </View>
        </View>
        <View
          numberOfSpaces={4}
          style={
            Array [
              Object {
                "width": 16,
              },
            ]
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "borderRadius": 8,
                  "height": 312,
                },
              ]
            }
          >
            <View>
              <Image
                imageHeight={288}
                imageWidth={192}
                source={
                  Object {
                    "uri": "https://storage.gra.cloud.ovh.net/v1/AUTH_688df1e25bd84a48a3804e7fa8938085/storage-pc-dev/thumbs/mediations/CDZQ",
                  }
                }
                style={
                  Array [
                    Object {
                      "borderTopLeftRadius": 8,
                      "borderTopRightRadius": 8,
                      "height": 288,
                      "width": 192,
                    },
                  ]
                }
                testID="offerTileImage"
              />
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#151515",
                      "borderBottomLeftRadius": 8,
                      "borderBottomRightRadius": 8,
                      "flexDirection": "row",
                      "height": 24,
                      "width": 192,
                    },
                  ]
                }
                width={192}
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexBasis": 0,
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "justifyContent": "center",
                        "paddingHorizontal": 4,
                      },
                    ]
                  }
                >
                  <Text
                    color="#ffffff"
                    style={
                      Array [
                        Object {
                          "color": "#ffffff",
                          "fontFamily": "Montserrat-SemiBold",
                          "fontSize": 12,
                          "lineHeight": 16,
                        },
                      ]
                    }
                    testID="categoryImageCaption"
                  >
                    Musique
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            imageWidth={192}
            style={
              Array [
                Object {
                  "marginTop": 8,
                  "maxWidth": 192,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#151515",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
            >
              I want something more
            </Text>
            <Text
              color="#626262"
              style={
                Array [
                  Object {
                    "color": "#626262",
                    "fontFamily": "Montserrat-SemiBold",
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                ]
              }
              testID="priceIsDuo"
            >
              28 €
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          numberOfSpaces={6}
          style={
            Array [
              Object {
                "width": 24,
              },
            ]
          }
        />
      </View>
    </View>
  </RCTScrollView>,
]
`;
